### DISCUSSION TITLE
Problem closing COM port with 1.5.8 (SerialUPDI, Windows)

### DISCUSSION THREAD

**Discussion by XFer012** [SpenceKonde/DxCore#463] [Q&A]
Hello,
after some time (about 1 year) I digged out my AVR128DB28 to do some testing, updated DxCore and... failed to upload a sketch.

OS: Windows 10 64 bit latest updates
IDE: Arduino 1.8.19
DxCore: 1.5.8 (updated from Boards Manager)
Serial-USB cable: CP2104 (works OK with ATTiny3224 and latest megaTinyCore)
Selected programmer: SerialUPDI (tried both "SLOW 57600" and 230400)

Problem:

```
Sketch uses 1616 bytes (1%) of program storage space. Maximum is 131072 bytes.
Global variables use 157 bytes (0%) of dynamic memory, leaving 16227 bytes for local variables. Maximum is 16384 bytes.
C:\Users\Fernando\AppData\Local\Arduino15\packages\megaTinyCore\tools\python3\3.7.2-post1/python3 -u C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8/tools/prog.py -t uart -u COM12 -b 57600 -d avr128db28 --fuses 6:0b00001100 7:0x00 8:0x00 -fC:\Temp\arduino-build-files/AVR_New_Sleep_PIT.ino.hex -a write -v 
SerialUPDI
UPDI programming for Arduino using a serial adapter
Based on pymcuprog, with significant modifications
By Quentin Bolsee and Spence Konde
Version 1.3.0.2 - Jun 2023
Using serial port COM12 at 57600 baud.
Target: avr128db28
Set fuses: ['6:0b00001100', '7:0x00', '8:0x00']
Action: write
File: C:\Temp\arduino-build-files/AVR_New_Sleep_PIT.ino.hex
pymcuprog.programmer - INFO - Setting up programming session for 'avr128db28'
pymcuprog.deviceinfo.deviceinfo - INFO - Looking for device avr128db28
pymcuprog.serialupdi.physical - INFO - Opening port 'COM12' at '57600' baud
pymcuprog.serialupdi.link - INFO - STCS 08 to 0x03
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.link - WARNING - UPDI init failed: Can't read CS register. likely wiring error.
pymcuprog.serialupdi.physical - INFO - Sending double break
pymcuprog.serialupdi.physical - INFO - Double-break sent. Retrying.
pymcuprog.serialupdi.physical - INFO - Opening port 'COM12' at '57600' baud
pymcuprog.serialupdi.link - INFO - STCS 08 to 0x03
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.link - INFO - UPDI init OK
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - Setting UPDI clock to 4 MHz
pymcuprog.serialupdi.link - INFO - STCS 03 to 0x09
pymcuprog.serialupdi.physical - INFO - Switching to '57600' baud
pymcuprog.serialupdi.application - WARNING - Cannot read SIB, hard reset...
pymcuprog.serialupdi.physical - INFO - Sending double break
pymcuprog.serialupdi.physical - INFO - Double-break sent. Retrying.
pymcuprog.serialupdi.physical - INFO - Opening port 'COM12' at '57600' baud
pymcuprog.serialupdi.application - INFO - SIB: 'AVR     P:2D:1-3M2 (A5.KV00D.0)'
pymcuprog.serialupdi.application - INFO - Device family ID: 'AVR'
pymcuprog.serialupdi.application - INFO - NVM interface: 'P:2'
pymcuprog.serialupdi.application - INFO - Debug interface: 'D:1'
pymcuprog.serialupdi.application - INFO - PDI oscillator: '3M2'
pymcuprog.serialupdi.application - INFO - Extra info: '(A5.KV00D.0)'
pymcuprog.serialupdi.application - INFO - Using 24-bit UPDI
pymcuprog.serialupdi.link - INFO - STCS 08 to 0x03
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.link - INFO - UPDI init OK
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.application - INFO - PDI revision = 0x03
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.application - INFO - Entering NVM programming mode
pymcuprog.serialupdi.link - INFO - LDCS from 0x07
pymcuprog.serialupdi.application - INFO - Apply reset
pymcuprog.serialupdi.link - INFO - STCS 59 to 0x08
pymcuprog.serialupdi.application - INFO - Release reset
pymcuprog.serialupdi.link - INFO - STCS 00 to 0x08
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.nvm - INFO - No specific initializer for this provider
Pinging device...
pymcuprog.programmer - INFO - Reading device ID...
pymcuprog.serialupdi.application - INFO - SIB: 'AVR     P:2D:1-3M2 (A5.KV00D.0)'
pymcuprog.serialupdi.application - INFO - Device family ID: 'AVR'
pymcuprog.serialupdi.application - INFO - NVM interface: 'P:2'
pymcuprog.serialupdi.application - INFO - Debug interface: 'D:1'
pymcuprog.serialupdi.application - INFO - PDI oscillator: '3M2'
pymcuprog.serialupdi.application - INFO - Extra info: '(A5.KV00D.0)'
pymcuprog.serialupdi.application - INFO - Using 24-bit UPDI
pymcuprog.serialupdi.link - INFO - STCS 08 to 0x03
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.link - INFO - UPDI init OK
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.application - INFO - PDI revision = 0x03
pymcuprog.serialupdi.link - INFO - LDCS from 0x0B
pymcuprog.serialupdi.application - INFO - Device ID = '1E970E' rev 'A5'
pymcuprog.nvm - INFO - Device ID: '1E970E'
pymcuprog.nvm - INFO - Device revision: 'A5'
pymcuprog.nvm - INFO - Device serial number: 'b'420501196000080701080119''
Ping response: 1E970E
Setting fuse 0x6=0xc
Writing literal values...
pymcuprog.programmer - INFO - Write...
pymcuprog.programmer - INFO - Writing 1 bytes of data to fuses...
pymcuprog.serialupdi.nvm - INFO - NVM EEPROM erase/write command
pymcuprog.serialupdi.nvm - INFO - Clear NVM command
pymcuprog.programmer - INFO - Write complete.
Verifying literal values...
pymcuprog.programmer - INFO - Reading 1 bytes from fuses...
pymcuprog.programmer - INFO - Verifying...
Action took 0.04s
Setting fuse 0x7=0x0
Writing literal values...
pymcuprog.programmer - INFO - Write...
pymcuprog.programmer - INFO - Writing 1 bytes of data to fuses...
pymcuprog.serialupdi.nvm - INFO - NVM EEPROM erase/write command
pymcuprog.serialupdi.nvm - INFO - Clear NVM command
pymcuprog.programmer - INFO - Write complete.
Verifying literal values...
pymcuprog.programmer - INFO - Reading 1 bytes from fuses...
pymcuprog.programmer - INFO - Verifying...
Action took 0.03s
Setting fuse 0x8=0x0
Writing literal values...
pymcuprog.programmer - INFO - Write...
pymcuprog.programmer - INFO - Writing 1 bytes of data to fuses...
pymcuprog.serialupdi.nvm - INFO - NVM EEPROM erase/write command
pymcuprog.serialupdi.nvm - INFO - Clear NVM command
pymcuprog.programmer - INFO - Write complete.
Verifying literal values...
pymcuprog.programmer - INFO - Reading 1 bytes from fuses...
pymcuprog.programmer - INFO - Verifying...
Action took 0.05s
Finished writing fuses.
Chip/Bulk erase,
Memory type eeprom is conditionally erased (depending upon EESAVE fuse setting)
Memory type flash is always erased
Memory type lockbits is always erased
...
pymcuprog.programmer - INFO - Erase...
pymcuprog.serialupdi.nvm - INFO - Chip erase using NVM CTRL
Traceback (most recent call last):
  File "C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8/tools/prog.py", line 287, in <module>
    main()
  File "C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8/tools/prog.py", line 129, in main
    return_code = pymcuprog_basic(args, fuses_dict)
  File "C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8/tools/prog.py", line 250, in pymcuprog_basic
    offset=0)
  File "C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8/tools/prog.py", line 139, in run_pymcu_action
    status = func(backend, *args, args_pymcu)
  File "C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\pymcuprog_main.py", line 189, in _action_erase
    backend.erase(args.memory, address=None)
  File "C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\backend.py", line 436, in erase
    self.programmer.erase(memory_name, address)
  File "C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\programmer.py", line 150, in erase
    self.device_model.erase(memory_info=memory_info, address=address)
  File "C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\nvmserialupdi.py", line 103, in erase
    self.avr.nvm.chip_erase()
  File "C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\serialupdi\nvm.py", line 241, in chip_erase
    if not self.wait_flash_ready():
  File "C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\serialupdi\nvm.py", line 59, in wait_flash_ready
    status = self.readwrite.read_byte(self.device.nvmctrl_address + constants.UPDI_NVMCTRL_STATUS)
  File "C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\serialupdi\readwrite.py", line 56, in read_byte
    return self.datalink.ld(address)
  File "C:\Users\Fernando\AppData\Local\Arduino15\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\serialupdi\link.py", line 410, in ld
    return self.updi_phy.receive(1)[0]
IndexError: bytearray index out of range
pymcuprog.serialupdi.physical - INFO - Closing port 'COM12'
the selected serial port pymcuprog.serialupdi.physical - INFO - Closing port 'COM12'
 does not exist or your board is not connected
```

I re-checked all wiring with my multimeter and everything seems fine; I had the test board stored away in a box and worked like a charm 1 year ago.
I know it could still have failed, maybe I will have to redo the board, but I cannot understand this part:

Pinging device...
pymcuprog.programmer - INFO - Reading device ID...
pymcuprog.serialupdi.application - INFO - SIB: 'AVR     P:2D:1-3M2 (A5.KV00D.0)'
pymcuprog.serialupdi.application - INFO - Device family ID: 'AVR'
pymcuprog.serialupdi.application - INFO - NVM interface: 'P:2'
pymcuprog.serialupdi.application - INFO - Debug interface: 'D:1'
pymcuprog.serialupdi.application - INFO - PDI oscillator: '3M2'
pymcuprog.serialupdi.application - INFO - Extra info: '(A5.KV00D.0)'

Doesn't it mean it was able to communicate with the MCU?

Best,
Fernando

**Comment by SpenceKonde** [comment]
Are you using the github version? I think I introduced a bug while working on something else.

**Reply by XFer012** [reply]
No, I'm using the one from Arduino Boards Manager (1.5.8)

**Reply by SpenceKonde** [reply]
looks like the timeout was on the edge of working; I'd already noticed the issue on mTC and changed it there but not here. You get callback errors instead of a message saying it got no response back because I don't know how to do that, all I know how to do is attempt to index the response and watch it blow up. I don't know how to work with errors properly in python to handle them more gracefully. Hell, I don't know how to handle errors gracefully in ANY language, since here in AVRland we run with -fno-exceptions. 

>the selected serial port pymcuprog.serialupdi.physical - INFO - Closing port 'COM12' does not exist or your board is not connected

This is a nonsense error and I have been trying to figure out where it came up with that that outrageous pants-on-fire lie (it's stuff we inherited from microchip's pymcuprog). Obviously the port exists, and obviously a board is connected to it because we just read out the NVM programming version, the on chip debugging version, the frequency of the PDI oscillator, and the die rev and an internal die ID from it, at minimum. Now if either of the claims of that error were true - then where did that info come from? Did I miss a product release announcement for a Ouija board on a chip or something?`*` The error is totally incorrect. It is the error shown basically at the end of every failed upload attempt. What happened in your case was that the time it allotted to the erase was insufficient (likely right on the edge of working), so when it tried to check the result, it got no reply, but only discovered this when it went to get the first byte of a bytestring with length 0.

That triggered the callstack dump. *but then for reasons I don't understand it is followed with that garbage line*

I just noticed that the method I ported over is not rigorous enough, and am working on what I hope will be sufficient

`*` I should get some if so - there's just no end to the fun you could have with this. With enough of them, we could summon, say, the souls of royal courts of the past and tell them what life was like now. And that would provide a route towards free energy - all we'd need is someone very prim and proper in a coffin that they can't roll over in, a shovel and a dynamo.; we could generate power from all these long dead and easily appalled individuals rolling over in their graves. It's sustainable, since the soul is immortal, though it's unclear whether we'd be able to locate enough such dead folks - but we can increase the supply of those and simultaneously decrease the demand for energy through a simple process.  And since political polarization is so bad, it would be easy to use generative AI to make untrue statements that horrify these restless dead such that they roll over in their graves and generate power. So this would just be a matter of rebalancing the portion of humans who have ever lived and the portion of humans who are now restless dead. And it just gets easier from there:  Since we would only need to account for cases where there were too many living energy-using humans and not enough restless dead to connect to dynamos, we would not need to confront the problem of converting dead people into living ones, only the opposite direction. That has been implemented effectively by many organizations (most efficiently by various military organizations) and can be considered a solved problem. Best of all, since we would be trying to maximize power generation, that would mean converting from sink to source the individuals who were most polarized and hence most toxic to the discussion, hence applying selective pressure to the human race to not be so damned opinionated on politics. Sustainable energy generation as well as conservation, and a long term nudge away from political polarization? What's not to like? I mean, other than the detail of converting living, energy consuming humans into restless-dead, and then using generative AI to tell outrageous and offensive lies to the dead using an array of ouija-on-a-chip devices, and turning cemeteries from monuments to our dearly departed into something like an underground windfarm.... Of course some people will be appalled by this idea - but as we could guarantee that those most horrified with this idea would be reliable generators, and become moreso as the technology was deployed more widely, it's likely that those folks would soon be contributing instead of criticizing.... 
("Demonstration: Bringing an observation to it's most Illogical conclusion, through a process of baseless assumptions and specious arguments")

**Reply by XFer012** [reply]
I love reading your replies! I should collect them in a book! :-D
I gather you are looking into it, maybe lengthening the timeout would suffice as a momentary fix?

**Comment by SpenceKonde** [comment]
Well, I don't think it was actually doing the timeout check loop correctly, period. Full stop. It looks like it would start the first passage through the loop 2 usb latency cycles later I think. The only thing that doesn't add up is that shouldn't be long enough. I know they usually highball the write delays, but not by that much - the datasheet "typical" of 70us for a word write is usually more like 60us. but 2 USB latency cycles is 4ms on a CH340, and it is supposed to take 11ms for a chip erase.....1/7th better than expected seems reasonable. 7/11ths better is implausible. So I still don't fully understand that, I'm not a good python programme - I missed CS201, Constrictor Snakes and Software Development
Try the version from github now.... I think I have trapped out the errors, and fixed a few other issues with the text displayed, 

To my profound dismay, searching the entire DxCore directory turned indicates that 

"the selected serial port does not exist or your board is not connected"

isn't in there. (and searches for things like "your board") does not turn up any results either. You know what *that* means: that garbage is coming from the IDE, and 

Somehow we're causing the IDE to print that message, and as you can see when it tries to print the port name, it's instead printing 
a message sent during normal operation.  pymcuprog.serialupdi.physical - INFO - Closing port 'COM12` - somehow it's getting that back from what it expects is the port name, but is actually a whole line of debug prints, effectively

Most likely, that variable is supposed to contain the name of the port, not a debug string.

**Reply by XFer012** [reply]
Before I horribly mess up my Arduino environment (1.8.19 Win64): since I already have 1.5.8 (from Boards Manager) installed, is there a safe way to install the Github version instead (that is, without having to offer flesh sacrifices to some divinity and without screwing up badly everything)?
Thanks!

**Reply by SpenceKonde** [reply]
First off - Blood sacrifices to the dark gods are mandatory regardless of what embedded development tasks you are hope to achieve. (and it's blood sacrifice, not flesh sacrifice. "Blood is the currency of the spirit world", not flesh. (attributed to the esteemed count in his discourse with B.  Stoker, one of the best known researchers in the field of )). So you're going to want to leave the altar and it's paraphenalia set up - yeah, i know it's kidna bulky, but it's one of the most important 

Unpack .zip package into a new location in someplace like the root of C, so you won't encounter windows permissions treating the directory wierd. 

Open that directory and create a folder inside called "portable". Open the IDE from that location (you now have two completely separate copies that are unaware of eachother.  Then close it. You'll see that a bunch of shit was created inside portable. 

Then either: 
Reopen it (the closing is really just a paranoia step), open board manager and install, exit, then go into C:\arduino-1.8.19-portable\portable\packages\DxCore\hardware\megaavr\versionnumber\tools- replace the updated .py files from the repo. This cannot by "synced" to the repo (if you notice, the directory tree is shaped differently). 

Or
Unpack zip and make portable folder and open+close IDE as above. 

Instead of reopening and installing via board manager, do the full manual install. 

First update the default toolchain - open the board manager json file URL in browser, Search for Azduino7, and hit next until you get the the one place with several instaces of that string in rapid succession, each associated with a platform and a URL. Download and extract archive from this location (suggest to use the platform that matches your computer, else nothing will work). Being careful that you're not putting the directory one layer too deep or not deep enough, open arduiinodir/Hardware/Tools/avr (the result is impossible to unpick if you mess up, because there are multiple folders in that tangled mess with the same name. Like theres's avr/bin, avr/avr/bin, avr/include and avr/avr/include). You will replace thousands of files. Do so. Open the IDE and verify that you can compile blink for some

Now open portable/sketchbook, create a folder named hardware in there, and sync the repo into that folder.

Finally, you need to open that folder, navigate to megaavr/tools/ManualPython.md open in text editor and follow the instructions there to get the python package we use.

Either way, you pick one path, and you stay the hell on it. Thou shalt not use board manager on any installation with an updated default toolchain. Thou shalt not install any cores but mTC, DxC, and ATTC 2.0.0-dev on this portable version (there are no concerns with libraries, just hardware packages). When there are new ATpacks, and there's an Azduino8, you can do the same thing to update the toolchain to that. We do not forsee a need to update the python package in the near future.

That's the setup I use for core development, except for compiling bootloaders. Bootloader binaries are a rather different animal, but it can be done similar steps, actually, - in some ways - but different IDE version and you delete almost all but one branch of the directory tree.

**Reply by XFer012** [reply]
WHOA! Not for the faint of heart...
Thanks, I'll try. That is, unless your are working on a new release which incorporates the fix. :-)

**Comment by SpenceKonde** [comment]
Well, I dunno about a release. 
I'm hoping to get avrdude 7.2 in but I said some undiplomatic things there and I think I just got flamed to a crisp over there, I'm a little afraid to go over there and read the thread. Thing is, it's C and so we're dependent on other people to compile the binaries for the fucking 7 stupid platforms that you need to support for Arduino (and you get shit like nobody's business if you don't support one, and the versions are not drop in compatible either - a few things get changed around in the most recent versions, which is excusable, they appear to have dealt with an enormous amount of technical debt which had built up in AVRdude - but that puts a solution where I can do a release with an incomplete set of binaries out of reach). I struggle to compile for x64 linux, and know nothing about toolchains or how to set up cross compiling environments or any of that kind of thing, so I don't have a prayer of achieving this myself. 

I found ways around this, cheap tricks, to make the AVR toolchain packages (do you think I ship any of the binaries I compiled? No way, they don't work! I use the ones from the last Arduino one, 3 years ago!), and I can only build for x64 linux anyway. Actually, I think the *compiler* works - it's the linker that's dead as a doornaill.That must be the case, since the build process precompiles a metric shitton of .o files that are used to make the compile time less of a nuisance - this is more than half of the compiler package size - 165 MB, about 6mb per part family that has it's own set of .a files. 

Oh, and maybe use the Azduino6 toolchain package and don't try to compile for  Ex (it doesn't work anyway). Just realized the windows version is hosed :-/ Again. Same story as always, avr/subdirs vs avr/avr/subdirs. I just can't seem to get it right. 

But in any event, i'd like to see that I made #463 better before I move forward with a release.

**Reply by SpenceKonde** [reply]
Actually no, Azduino7 is fine. I was thrown off while checking directory stats on it.

**Comment by XFer012** [comment]
Hello Spence,

I did as instructed, but the new prog.py throws a syntax error:

```
Sketch uses 1624 bytes (1%) of program storage space. Maximum is 131072 bytes.
Global variables use 157 bytes (0%) of dynamic memory, leaving 16227 bytes for local variables. Maximum is 16384 bytes.
F:\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\megaTinyCore\tools\python3\3.7.2-post1/python3 -u F:\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8/tools/prog.py -t uart -u COM3 -b 230400 -d avr128db28 --fuses 6:0b00010100 7:0x00 8:0x00 -fb:\temp\arduino_build_895862/AVR_New_Sleep_PIT.ino.hex -a write -v 
  File "F:\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8/tools/prog.py", line 101
An error occurred while uploading the sketch
    if args.action in ("read", "write") and args.filename == "" and (args.fuses == "" or args.action = "read") and not args.fuses_print:
                                                                                                     ^
SyntaxError: invalid syntax

```

Update: fixing the .py (args.action == "read"), the upload went FINE by using a CH340.
Unfortunately I did not carry with me the CP2102N I used when reporting the problem in the first place. I should have a chance of retrieving it in a couple days.

**Comment by XFer012** [comment]
Unfortunately, not even the updated prog.py works with the other USB-Serial cable:

```
Sketch uses 1624 bytes (1%) of program storage space. Maximum is 131072 bytes.
Global variables use 157 bytes (0%) of dynamic memory, leaving 16227 bytes for local variables. Maximum is 16384 bytes.
J:\Progetti\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\megaTinyCore\tools\python3\3.7.2-post1/python3 -u J:\Progetti\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8/tools/prog.py -t uart -u COM12 -b 230400 -d avr128db28 --fuses 6:0b00010100 7:0x00 8:0x00 -fc:\temp\arduino_build_667054/AVR_New_Sleep_PIT.ino.hex -a write -v 
SerialUPDI
UPDI programming for Arduino using a serial adapter
Based on pymcuprog, with significant modifications
By Quentin Bolsee and Spence Konde
Version 1.3.0.3 - Jul 2023
Using serial port COM12 at 230400 baud.
Target: avr128db28
Set fuses: ['6:0b00010100', '7:0x00', '8:0x00']
Action: write
File: c:\temp\arduino_build_667054/AVR_New_Sleep_PIT.ino.hex
pymcuprog.programmer - INFO - Setting up programming session for 'avr128db28'
pymcuprog.deviceinfo.deviceinfo - INFO - Looking for device avr128db28
pymcuprog.serialupdi.physical - INFO - Opening port 'COM12' at '115200' baud
pymcuprog.serialupdi.link - INFO - STCS 08 to 0x03
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.link - INFO - UPDI init OK
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - Setting UPDI clock to 8 MHz
pymcuprog.serialupdi.link - INFO - STCS 02 to 0x09
pymcuprog.serialupdi.physical - INFO - Switching to '230400' baud
pymcuprog.serialupdi.application - INFO - SIB: 'AVR     P:2D:1-3M2 (A5.KV00D.0)'
pymcuprog.serialupdi.application - INFO - Device family ID: 'AVR'
pymcuprog.serialupdi.application - INFO - NVM interface: 'P:2'
pymcuprog.serialupdi.application - INFO - Debug interface: 'D:1'
pymcuprog.serialupdi.application - INFO - PDI oscillator: '3M2'
pymcuprog.serialupdi.application - INFO - Extra info: '(A5.KV00D.0)'
pymcuprog.serialupdi.application - INFO - Using 24-bit UPDI
pymcuprog.serialupdi.link - INFO - STCS 08 to 0x03
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.link - WARNING - UPDI init failed: Can't read CS register. likely wiring error.
pymcuprog.serialupdi.physical - INFO - Sending double break
pymcuprog.serialupdi.physical - INFO - Double-break sent. Retrying.
pymcuprog.serialupdi.physical - INFO - Opening port 'COM12' at '115200' baud
pymcuprog.serialupdi.link - INFO - STCS 08 to 0x03
pymcuprog.serialupdi.link - INFO - STCS 06 to 0x02
pymcuprog.serialupdi.link - INFO - LDCS from 0x00
pymcuprog.serialupdi.link - WARNING - UPDI init failed: Can't read CS register. likely wiring error.
Traceback (most recent call last):
  File "J:\Progetti\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8/tools/prog.py", line 287, in <module>
    main()
  File "J:\Progetti\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8/tools/prog.py", line 129, in main
    return_code = pymcuprog_basic(args, fuses_dict)
  File "J:\Progetti\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8/tools/prog.py", line 202, in pymcuprog_basic
    args_start)
  File "J:\Progetti\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\pymcuprog_main.py", line 545, in _start_session
    backend.start_session(sessionconfig)
  File "J:\Progetti\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\backend.py", line 362, in start_session
    sessionconfig.interface_speed)
  File "J:\Progetti\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\programmer.py", line 83, in setup_device
    options=self.options)
  File "J:\Progetti\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\nvm.py", line 42, in get_nvm_access_provider
    accessprovider = NvmAccessProviderSerial(transport, device_info, baud=frequency, options=options)
  File "J:\Progetti\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\nvmserialupdi.py", line 57, in __init__
    self.avr.read_device_info()
  File "J:\Progetti\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\serialupdi\application.py", line 114, in read_device_info
    datalink.init_datalink()
  File "J:\Progetti\Arduino\_ARDUINO_PORTABLE\arduino-1.8.19\portable\packages\DxCore\hardware\megaavr\1.5.8\tools\libs\pymcuprog\serialupdi\link.py", line 44, in init_datalink
    raise PymcuprogError("UPDI initialisation failed")
pymcuprog.pymcuprog_errors.PymcuprogError: UPDI initialisation failed
pymcuprog.serialupdi.physical - INFO - Closing port 'COM12'
the selected serial port pymcuprog.serialupdi.physical - INFO - Closing port 'COM12'
 does not exist or your board is not connected

```

**Comment by XFer012** [comment]
@SpenceKonde should I convert this into an issue?
